### PR TITLE
chore: replace TF Cloud implementation with TF CLI

### PR DESCRIPTION
**Pull request by DanielMSchmidt**
This will allow us to integrate better with TFC and TFE without additional work being spend on polishing this part.

## TODO

- [x] Check if this solves https://github.com/hashicorp/terraform-cdk/issues/490
- [ ] Check if this solves https://github.com/hashicorp/terraform-cdk/issues/489

### CODE REVIEW

**Comment by ansgarm** [comment]
Okay, so #490 does not quite work yet with this PR and requires a follow-up:
Screenshot 2022-11-22 at 11 56 11

**Comment by michaellzc** [comment]
Hey, this seems like a better approach otherwise `cdktf` will always have to play catch-up with TFC/TFE, e.g. https://github.com/hashicorp/terraform-cdk/pull/2243.

but... when I tried out this branch, and got this error when running `cdktf destroy`

```
[2022-11-25T19:30:29.198] [ERROR] default - ╷
│ Error: Saving a generated plan is currently not supported
│
│ Terraform Cloud does not support saving the generated execution plan
output  ╷
        │ Error: Saving a generated plan is currently not supported
        │
        │ Terraform Cloud does not support saving the generated execution plan
        │ locally at this time.
```

Apparently, TFC/TFE does not support saving the plan locally through `terraform` cli when remote execution is enabled, and the raw plan is only exposed through their API. (TIL we just migrated to TFC not that long ago)

Therefore, it seems almost impossible to get the state where `cdktf` has zero knowledge of how TFC/TFE works, and always requires some special handling.

Is such a huge discrepancy between normal terraform behaviour and TFC/TFE something on the terraform and cdktf team radar? What is the plan for cdktf to address it?

**Comment by ansgarm** [comment]
Hi @michaellzc 👋

Wow, you are quite the adventurer by daring to test this draft PR 😁👏

Yeah, I noticed that as well when I took over Daniel's work on this. In the long term, we're working with the team working on the Terraform CLI on a solution that allows us to close this gap (I don't know yet whether they'll support saving a TFC/E remote plan in a file that can be applied or whether a different solution will be preferred).

However, as we need to support current Terraform CLI versions as well, our current approach is to use the Terraform CLI in an interactive way (via `node-pty`). So I'm currently working on switching the implementation from `cdktf deploy = terraform plan -out <plan> && terraform apply <plan>` to `cdktf deploy = terraform apply` (and relaying the "Do you want to apply? yes / no"-question). This will work for all cases but still has uncertainties that we need to test.

**Comment by team-tf-cdk** [comment]
This PR exceeds the recommended size of 1000 lines. Please make sure you are NOT addressing multiple issues with one PR. Note this PR might be rejected due to its size.

**Comment by team-tf-cdk** [comment]
This PR exceeds the recommended size of 1000 lines. Please make sure you are NOT addressing multiple issues with one PR. Note this PR might be rejected due to its size.

**Comment by team-tf-cdk** [comment]
This PR exceeds the recommended size of 1000 lines. Please make sure you are NOT addressing multiple issues with one PR. Note this PR might be rejected due to its size.

**Comment by team-tf-cdk** [comment]
This PR exceeds the recommended size of 1000 lines. Please make sure you are NOT addressing multiple issues with one PR. Note this PR might be rejected due to its size.

File: packages/@cdktf/cli-core/src/lib/models/terraform-cli.ts [hunk 1]
```diff
@@ -49,6 +61,27 @@ export class TerraformCli implements Terraform {
       args.push("-upgrade");
     }
 
+    // TODO: do we need to run this if we use the TF CLI to run in TFC?
```
Comment (mutahhir) [inline packages/@cdktf/cli-core/src/lib/models/terraform-cli.ts:64]: Yes, if we don't do this, Remote execution starts failing with:

image

Comment (ansgarm) [reply packages/@cdktf/cli-core/src/lib/models/terraform-cli.ts:64]: Damn.. We currently have no way to know whether something runs in TFC with remote execution 😬

**Comment by team-tf-cdk** [comment]
This PR exceeds the recommended size of 1000 lines. Please make sure you are NOT addressing multiple issues with one PR. Note this PR might be rejected due to its size.

File: packages/@cdktf/cli-core/src/lib/models/terraform-cli.ts (hunk 1 above)
Comment (mutahhir) [reply packages/@cdktf/cli-core/src/lib/models/terraform-cli.ts:69]: I haven't seen a problem with local exec using this.

Comment (ansgarm) [reply packages/@cdktf/cli-core/src/lib/models/terraform-cli.ts:69]: Nope, but it downloads three provider binaries, so I do notice this on my slow German internet connection 😅

Comment (ansgarm) [reply packages/@cdktf/cli-core/src/lib/models/terraform-cli.ts:69]: Or at least something network heavy seems to be happening – I'd love to be wrong here 😄

**Comment by team-tf-cdk** [comment]
This PR exceeds the recommended size of 1000 lines. Please make sure you are NOT addressing multiple issues with one PR. Note this PR might be rejected due to its size.

**Comment by team-tf-cdk** [comment]
This PR exceeds the recommended size of 1000 lines. Please make sure you are NOT addressing multiple issues with one PR. Note this PR might be rejected due to its size.

**Comment by team-tf-cdk** [comment]
This PR exceeds the recommended size of 1000 lines. Please make sure you are NOT addressing multiple issues with one PR. Note this PR might be rejected due to its size.

**Comment by team-tf-cdk** [comment]
This PR exceeds the recommended size of 1000 lines. Please make sure you are NOT addressing multiple issues with one PR. Note this PR might be rejected due to its size.

**Comment by team-tf-cdk** [comment]
This PR exceeds the recommended size of 1000 lines. Please make sure you are NOT addressing multiple issues with one PR. Note this PR might be rejected due to its size.

**Comment by team-tf-cdk** [comment]
This PR exceeds the recommended size of 1000 lines. Please make sure you are NOT addressing multiple issues with one PR. Note this PR might be rejected due to its size.

**Comment by team-tf-cdk** [comment]
This PR exceeds the recommended size of 1000 lines. Please make sure you are NOT addressing multiple issues with one PR. Note this PR might be rejected due to its size.

**Comment by team-tf-cdk** [comment]
This PR exceeds the recommended size of 1000 lines. Please make sure you are NOT addressing multiple issues with one PR. Note this PR might be rejected due to its size.

**Comment by team-tf-cdk** [comment]
This PR exceeds the recommended size of 1000 lines. Please make sure you are NOT addressing multiple issues with one PR. Note this PR might be rejected due to its size.

**Comment by team-tf-cdk** [comment]
This PR exceeds the recommended size of 1000 lines. Please make sure you are NOT addressing multiple issues with one PR. Note this PR might be rejected due to its size.

**Comment by team-tf-cdk** [comment]
This PR exceeds the recommended size of 1000 lines. Please make sure you are NOT addressing multiple issues with one PR. Note this PR might be rejected due to its size.

**Comment by team-tf-cdk** [comment]
This PR exceeds the recommended size of 1000 lines. Please make sure you are NOT addressing multiple issues with one PR. Note this PR might be rejected due to its size.

**Comment by team-tf-cdk** [comment]
This PR exceeds the recommended size of 1000 lines. Please make sure you are NOT addressing multiple issues with one PR. Note this PR might be rejected due to its size.

**Comment by team-tf-cdk** [comment]
This PR exceeds the recommended size of 1000 lines. Please make sure you are NOT addressing multiple issues with one PR. Note this PR might be rejected due to its size.

**Comment by team-tf-cdk** [comment]
This PR exceeds the recommended size of 1000 lines. Please make sure you are NOT addressing multiple issues with one PR. Note this PR might be rejected due to its size.

**Comment by team-tf-cdk** [comment]
This PR exceeds the recommended size of 1000 lines. Please make sure you are NOT addressing multiple issues with one PR. Note this PR might be rejected due to its size.

**Comment by team-tf-cdk** [comment]
This PR exceeds the recommended size of 1000 lines. Please make sure you are NOT addressing multiple issues with one PR. Note this PR might be rejected due to its size.

**Comment by team-tf-cdk** [comment]
This PR exceeds the recommended size of 1000 lines. Please make sure you are NOT addressing multiple issues with one PR. Note this PR might be rejected due to its size.

**Comment by team-tf-cdk** [comment]
This PR exceeds the recommended size of 1000 lines. Please make sure you are NOT addressing multiple issues with one PR. Note this PR might be rejected due to its size.

**Comment by team-tf-cdk** [comment]
This PR exceeds the recommended size of 1000 lines. Please make sure you are NOT addressing multiple issues with one PR. Note this PR might be rejected due to its size.

**Comment by team-tf-cdk** [comment]
This PR exceeds the recommended size of 1000 lines. Please make sure you are NOT addressing multiple issues with one PR. Note this PR might be rejected due to its size.

**Comment by team-tf-cdk** [comment]
This PR exceeds the recommended size of 1000 lines. Please make sure you are NOT addressing multiple issues with one PR. Note this PR might be rejected due to its size.

File: packages/@cdktf/cli-core/src/lib/models/terraform-cli.ts (hunk 1 above)
Comment (ansgarm) [reply packages/@cdktf/cli-core/src/lib/models/terraform-cli.ts]: fixed 👍 (this was also caught by multiple integration tests, after we got them running)

**Comment by team-tf-cdk** [comment]
This PR exceeds the recommended size of 1000 lines. Please make sure you are NOT addressing multiple issues with one PR. Note this PR might be rejected due to its size.

**Comment by team-tf-cdk** [comment]
This PR exceeds the recommended size of 1000 lines. Please make sure you are NOT addressing multiple issues with one PR. Note this PR might be rejected due to its size.

**Comment by team-tf-cdk** [comment]
This PR exceeds the recommended size of 1000 lines. Please make sure you are NOT addressing multiple issues with one PR. Note this PR might be rejected due to its size.

**Comment by team-tf-cdk** [comment]
This PR exceeds the recommended size of 1000 lines. Please make sure you are NOT addressing multiple issues with one PR. Note this PR might be rejected due to its size.

**Review by ansgarm — APPROVED** [review]

**Pull request merged**